### ISSUE TITLE
Handle cases where ceph command blocks indefinitely

### ISSUE THREAD

**Issue by turettn** [ceph/ceph-csi#1216]
# Describe the bug #

If the `ceph` command hangs indefinitely, k8s will eventually give up.  ceph-csi will not, and any subsequent attempt to do anything to that volume will result in an "operation in progress" error.

We're seeing a weird bug on the ceph side where a bunch of ceph commands (e.g. `ceph fs subvolume getpath`) blocks.  We're working on that separately, but it exposed this issue with ceph-csi.

# Environment details #

- Image/version of Ceph CSI driver : 2.1.2
- Helm chart version : N/A
- Kernel version : 5.5.1
- Mounter used for mounting PVC (for cephfs its `fuse` or `kernel`. for rbd its
  `krbd` or `rbd-nbd`) : kernel
- Kubernetes cluster version : 1.18.0
- Ceph cluster version : 14.2.8

# Steps to reproduce #

1.  Get ceph into a bad state.  I haven't managed to make this reproducible yet.  For demo purposes, swap the ceph binary with an infinite loop.
2.  Create a pod that uses a currently unused PV.

# Actual results #

* k8s sends a nodeStageVolume request
* The volumeLock for that volumeId is acquired.  It is released when the function returns.
* A bunch of ExecCommands occur, one of which blocks indefinitely.  The function does not return, and the lock is still held.
* k8s times out
* Another pod requests the same PVC, or the original pod is cancelled, resulting in another nodeStageVolume or nodeUnstageVolume.  This request fails with VolumeOperationAlreadyExists, and will continue to do so until you bounce the csi driver.

# Expected behavior #

k8s times out, and assumes the operation was not successful.  ceph-csi needs to do something similar.

Simply putting a [timeout](https://medium.com/@vCabbage/go-timeout-commands-with-os-exec-commandcontext-ba0c861ed738) on the ExecCommand call seems like the right answer to me.  k8s will timeout, or ceph-csi will and then return an error.

Either way, subsequent operations (e.g. unstageVolume, or stageVolume after ceph is healthy again) will not be blocked by the acquired mutex.

# Additional context #

This isn't the bug I'm having, but is a good example of something that would cause the problem: [Ceph bug #41903](https://tracker.ceph.com/issues/41903)

Golang is definitely not a language I touch frequently, but if people agree that this is a reasonable approach, I will happily try to piece together a PR to add timeouts.

**Comment by nixpanic** [comment]
We are removing all calls to `ceph`, `rados` and `rbd` commands in favour of ceph/go-ceph API calls.

That said, there is no timeout in the Ceph libraries that can be used for this. ceph/go-ceph#165 has been filed for a similar request.

**Comment by Madhu-1** [comment]
I tried to do add a timeout for CLI commands long back https://github.com/ceph/ceph-csi/pull/194 but we closed that one as it can cause issues. @nixpanic is it something we can request from ceph libraries?

is there a way to identify which command is struck in go-ceph now?

**Comment by nixpanic** [comment]
> is there a way to identify which command is struck in go-ceph now?

There is no way at the moment. We could consider adding something optionally to the `rados.IOContext` or somewhere else. I imagine a member added to one of the types that can be used to construct a `context.Context` when calling out to the Ceph libraries. @phlogistonjohn and @ansiwen, what do you think?

**Comment by phlogistonjohn** [comment]
There are some ceph configuration values with timeouts that can be experimented with and tweaked.

Additionally, I would suggest a mechanism for monitoring long running calls using a goroutine. Perhaps something like:
```go
monitor := NewCephCallMonitor(ctx)
defer monitor.Done()
// do an api call with go-ceph
```
Or for more complex multi call function
```go
monitor := NewCephCallMonitor(ctx)
defer monitor.Done()
monitor.Update("cephApiCallOne")
// do cephApiCallOne
montor.Update("cephApiCallTwo")
// do cephApiCallTwo
```

This approach would work for both go-ceph calls or cli calls. It would basically act as a supplement to inline logging and would log warnings if the api calls exceeded the time out values in the context. For the complex functions, the Update calls would indicate what api is being used and reset the timeout each time its called indicating progress has been made.

**Comment by nixpanic** [comment]
This is actually a duplicate of #694.
